scripts/irods/test/test_pam_password.py

```diff
@@ -0,0 +1,340 @@
+                                ['iinit'], 'STDOUT', 'PAM password', input=f'{self.auth_session.password}\n')
+                            self.auth_session.assert_icommand(["ils"], 'STDOUT', self.auth_session.session_collection)
+
+        finally:
+            self.auth_session.environment_file_contents = auth_session_env_backup
+            self.admin.environment_file_contents = admin_session_env_backup
+
+            IrodsController().restart()
+
+            self.admin.assert_icommand(
+                ['iadmin', 'set_grid_configuration', self.configuration_namespace, _option_name, original_config])
+
+    def test_invalid_password_max_time(self):
+        self.do_test_invalid_password_time_configurations('password_max_time')
+
+    def test_invalid_password_min_time(self):
+        self.do_test_invalid_password_time_configurations('password_min_time')
+
+    def test_password_max_time_less_than_password_min_time_makes_ttl_constraints_unsatisfiable(self):
+        min_time_option_name = 'password_min_time'
+        max_time_option_name = 'password_max_time'
+
+        # Stash away the original configuration for later...
+        original_min_time = self.admin.assert_icommand(
+            ['iadmin', 'get_grid_configuration', self.configuration_namespace, min_time_option_name], 'STDOUT')[1].strip()
+
+        original_max_time = self.admin.assert_icommand(
+            ['iadmin', 'get_grid_configuration', self.configuration_namespace, max_time_option_name], 'STDOUT')[1].strip()
+
+        IrodsController().stop()
+
+        auth_session_env_backup = copy.deepcopy(self.auth_session.environment_file_contents)
+        admin_session_env_backup = copy.deepcopy(self.admin.environment_file_contents)
+        try:
+            service_account_environment_file_path = os.path.join(
+                os.path.expanduser('~'), '.irods', 'irods_environment.json')
+            with lib.file_backed_up(service_account_environment_file_path):
+                client_update = test_configurations.make_dict_for_ssl_client_environment(
+                    self.server_key_path, self.chain_pem_path, self.dhparams_pem_path)
+
+                lib.update_json_file_from_dict(service_account_environment_file_path, client_update)
+
+                self.admin.environment_file_contents.update(client_update)
+
+                client_update['irods_authentication_scheme'] = self.authentication_scheme
+                self.auth_session.environment_file_contents.update(client_update)
+
+                with temporary_core_file() as core:
+                    core.add_rule(test_configurations.get_pep_for_ssl(self.plugin_name))
+
+                    IrodsController().start()
+
+                    # Make sure the settings applied correctly...
+                    self.admin.assert_icommand(
+                        ['iadmin', 'get_grid_configuration', self.configuration_namespace, max_time_option_name], 'STDOUT')
+
+                    for base_ttl_in_hours in [2, 336]:
```

Comment:
Are 2 and 336 special?

Reply:
Yes and no :)

2 is the lowest possible value that we can test with (that is, 2 - 1 = 1, which is the smallest TTL that we can specify with `iinit`). 336 is 1209600 seconds, which is the default `max_password_time`, so it seemed like these were appropriate as boundary tests.

I added a comment explaining why these were chosen for the subTests.

Reply:
Ah, that makes sense.